### PR TITLE
Sentry setup

### PR DESCRIPTION
**Pull request by snopoke**

### CODE REVIEW

File: corehq/middleware.py [hunk 1]
```diff
@@ -167,3 +168,30 @@ def process_response(self, request, response):
...
         return getattr(response, '_always_allow_browser_caching', False)
+
+
+class SentryContextMiddleware(object):
+    """Add details to Sentry context.
+    Should be placed after 'corehq.apps.users.middleware.UsersMiddleware'
+    """
+    def __init__(self):
+        try:
+            from raven.contrib.django.raven_compat.models import client
```
Comment (dimagimon) [inline corehq/middleware.py]: F401 'client' imported but unused

**Comment by snopoke** [comment]
I think this is GTG. It's been on staging for a while.

File: settings.py [hunk 2]
```diff
@@ -1937,3 +1953,30 @@
 REST_FRAMEWORK = {
     'DATETIME_FORMAT': '%Y-%m-%dT%H:%M:%S.%fZ'
 }
+
+if SENTRY_PUBLIC_KEY and SENTRY_PRIVATE_KEY and SENTRY_PROJECT_ID:
+    import os
```
Comment (benrudolph) [inline settings.py]: nit this is already imported

**Review by benrudolph — APPROVED** [review]
sweet looking forward to this change

File: settings.py (hunk 2 above)
Comment (snopoke) [inline settings.py]: bc60c95

**Review by benrudolph — APPROVED** [review]

**Review by mkangia — APPROVED** [review]

**Comment by snopoke** [comment]
@benrudolph made a few changes: https://github.com/dimagi/commcare-hq/pull/14808/files/bc60c95cd88b58e489cc669b24c65985268f6fe4..c4062d852cb7e7f351c04ba07a85343f4d6614ea

Most notably trying to get human readable release names rather than git hashes.

File: settingshelper.py [hunk 3]
```diff
@@ -199,3 +200,37 @@ def fix_logger_obfuscation(fix_logger_obfuscation_, logging_config):
...
+        return
+
+    try:
+        from raven import breadcrumbs, fetch_git_sha
+    except ImportError:
+        return
+
+    release_dir = base_dir.split('/')[-1]
+    if re.match('\d{4}-\d{2}-\d{2}_\d{2}.\d{2}', release_dir):
+        release = "{}_{}".format(server_env, release_dir)
```
Comment (benrudolph) [inline settingshelper.py]: nice how does this show up on sentry? could be cool to format this as a link to the github tag (should match the release dir date)

**Review by benrudolph — APPROVED** [review]

File: settingshelper.py (hunk 3 above)
Comment (snopoke) [inline settingshelper.py]: It appears in the error tags:

image

And on the releases page:

image

Making it the URL to the tag would be long (https://github.com/dimagi/commcare-hq/releases/tag/2017-02-16_09.43-staging-deploy) but I think it'll change the format to match the tag name.

Comment (snopoke) [inline settingshelper.py]: hmm, tags format differes if it's an offline deploy

Comment (benrudolph) [inline settingshelper.py]: we could change the offline tag format. i'm not entirely sure it's necessary to differentiate the two

Comment (benrudolph) [inline settingshelper.py]: (i'm also happy to merge this now if you'd like and you can tinker with it later)

**Pull request merged**